pkg/controller/azvolume.go

```diff
@@ -78,6 +78,7 @@ var allowedTargetVolumeStates = map[string][]string{
 	string(v1alpha1.VolumeUpdating):         {string(v1alpha1.VolumeUpdated), string(v1alpha1.VolumeUpdateFailed)},
 	string(v1alpha1.VolumeCreated):          {string(v1alpha1.VolumeUpdating), string(v1alpha1.VolumeDeleting)},
 	string(v1alpha1.VolumeUpdated):          {string(v1alpha1.VolumeUpdating), string(v1alpha1.VolumeDeleting)},
+	string(v1alpha1.VolumeDeletionFailed):   {},
```

Comment:
Shouldn't the target states be `{string(v1alpha1.VolumeDeleting)}`? After, an empty set would still prevent the transition.

Reply:
BTW, `allowedTargetAttachmentStates` has a case for this: https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/be7a2db911cf6d8f7a8ce984c6ac41294a712a61/pkg/controller/attach_detach.go#L79

Reply:
A failed volume deletion should not be retried unless `crdProvisioner` triggers the retry by resetting the state.

Reply:
> BTW, `allowedTargetAttachmentStates` has a case for this:
> 
> https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/be7a2db911cf6d8f7a8ce984c6ac41294a712a61/pkg/controller/attach_detach.go#L79

I think this should be amended as well if we are not allowing controllers to do retries on failed creation / deletion / attachment / detachment on its own.

Reply:
> > BTW, `allowedTargetAttachmentStates` has a case for thi
> > https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/be7a2db911cf6d8f7a8ce984c6ac41294a712a61/pkg/controller/attach_detach.go#L79
> 
> I think this should be amended as well if we are not allowing controllers to do retries on failed creation / deletion / attachment / detachment on its own.

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/be7a2db911cf6d8f7a8ce984c6ac41294a712a61/pkg/controller/attach_detach.go#L78-L79

Reply:
Hmm. I see. Yes, this makes sense and you're right in that the controller shouldn't try the failed operation on its own.

Reply:
> I think this should be amended as well if we are not allowing controllers to do retries on failed creation / deletion / attachment / detachment on its own.

Looking at the [e2e multi-az test results](https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/actions/runs/1293240816), the test failed because it timed out while delete the driver. The reason appears to be because of a couple of leaked replica attachments in the `AttachmentFailed` state. Without the state transition from the failed states, self-initiated retries are prevented and there doesn't appear to be anything in the replica controller that would initiate an explicit retry. When the replica is later marked for deletion (i.e. when the primary attachment is deleted), this also fails because of the inability to transition from the failed. The leaked replicas (and subsequent inability to delete them) appears to prevent the driver's namespace from being uninstalled which is what usually causes the driver uninstall to hang.

Reply:
So it seems like we should allow self-initiated retries for target states as follows:
```
VolumeCreatingFailed --> VolumeDeleting
VolumeDeletionFailed --> VolumeDeleting
AttachmentFailed --> Detaching
DetachmentFailed --> Detaching
```

Reply:
I wonder if we should delete replica attachment if it failed attaching in `attach_detach_controller` so that `replica_controller` can view the deletion event and create a new replacement replica attachment on a different or same node based on current cluster situation.

Reply:
> So it seems like we should allow self-initiated retries for target states as follows:
> 
> ```
> VolumeCreatingFailed --> VolumeDeleting
> VolumeDeletionFailed --> VolumeDeleting
> AttachmentFailed --> Detaching
> DetachmentFailed --> Detaching
> ```
> 
> And we would need a special retry logic for failed attachment in replica controller.

And instead of adding target states like this (as this will result in undesired retries as well), we could add a new state called `ForceDetach` and allow target state update from `DetachmentFailed` to `ForceDetach`, and add a new logic in `replica controller` that updates replica's state to `ForceDetach` to retrigger the detachment (something similar to what we do in `crdProvisioner.

Reply:
For a primary attachment, `CrdProvisioner.PublishVolume` should delete the `AzVolumeAttachment` on failure. We have similar logic in `CrdProvisioner.CreateVolume` for a failed create operation.

Similary, `replica_controller` could handle the failed attach as you propose.

Reply:
> And instead of adding target states like this (as this will result in undesired retries as well), we could add a new state called `ForceDetach` and allow target state update from `DetachmentFailed` to `ForceDetach`, and add a new logic in `replica controller` that updates replica's state to `ForceDetach` to retrigger the detachment (something similar to what we do in `crdProvisioner`.

Would `ForceDetach` actually try another detach operation through `CloudProvisioner` or simply deleted the `AzVolumeAttachment` CRI?

Reply:
It should try another detach operation through `cloudProvisioner` as otherwise, we might end up not being able to delete the volume.

And this is no different from how `crdProvisioner` retriggers failed volume detachment by updating state to `Attached`. We simply cannot do the same for `replica controller` because we have no guarantee that `AzVolumeAttachment`'s state was `Attached` prior to detachment attempt unlike `crdProvisioner`.